core/extensions.rs

```diff
@@ -389,44 +392,43 @@ macro_rules! extension {
       };
 
       let state_fn: fn(&mut $crate::OpState, Config $( <  $( $param ),+ > )? ) = $(  $state_fn  )?;
-      $ext.state(move |state: &mut $crate::OpState| {
+      $ext.opstate_fn = Some(Box::new(move |state: &mut $crate::OpState| {
         state_fn(state, config);
-      });
+      }));
     }
   };
 
   (! __config__ $ext:ident $( parameters = [ $( $param:ident : $type:ident ),+ ] )? $( state_fn = $state_fn:expr )? ) => {
-    $(  $ext.state($state_fn);  )?
+    $( $ext.opstate_fn = Some(Box::new($state_fn)); )?
   };
 
   (! __ops__ $ext:ident __eot__) => {
   };
 
   (! __ops__ $ext:ident $ops_symbol:ident __eot__) => {
-    $ext.ops($ops_symbol())
+    $ext.ops.to_mut().extend($ops_symbol())
   };
 
   (! __ops__ $ext:ident $ops_symbol:ident < $ops_param:ident > __eot__) => {
-    $ext.ops($ops_symbol::<$ops_param>())
+    $ext.ops.to_mut().extend($ops_symbol::<$ops_param>())
   };
 }
 
-#[derive(Default)]
 pub struct Extension {
-  pub(crate) name: &'static str,
-  js_files: Vec<ExtensionFileSource>,
-  esm_files: Vec<ExtensionFileSource>,
-  esm_entry_point: Option<&'static str>,
-  ops: Option<Vec<OpDecl>>,
-  opstate_fn: Option<Box<OpStateFn>>,
-  middleware_fn: Option<Box<OpMiddlewareFn>>,
-  event_loop_middleware: Option<Box<EventLoopMiddlewareFn>>,
-  global_template_middleware: Option<Box<GlobalTemplateMiddlewareFn>>,
-  global_object_middleware: Option<Box<GlobalObjectMiddlewareFn>>,
-  external_references: Option<Vec<v8::ExternalReference<'static>>>,
-  initialized: bool,
-  enabled: bool,
-  deps: Option<&'static [&'static str]>,
+  pub name: &'static str,
+  pub deps: &'static [&'static str],
+  pub js_files: Cow<'static, [ExtensionFileSource]>,
+  pub esm_files: Cow<'static, [ExtensionFileSource]>,
+  pub esm_entry_point: Option<&'static str>,
+  pub ops: Cow<'static, [OpDecl]>,
+  pub external_references: Cow<'static, [v8::ExternalReference<'static>]>,
+  pub opstate_fn: Option<Box<OpStateFn>>,
+  pub middleware_fn: Option<Box<OpMiddlewareFn>>,
+  pub enabled: bool,
+  pub js_enabled: bool,
```

Comment:
`js_enabled` might need a better name and/or some docs. Perhaps `will_contribute_js_to_runtime`? It isn't visible to anything but user code.

Reply:
Might not need this after all, maybe I can just do `ext.esm_files = Cow::Borrowed(&[])` etc. in place of `js_enabled = false`.